blink/blinkenlights.c

```diff
@@ -3078,6 +3103,87 @@ static void OnBaseMemSizeService(void) {
   Put16(m->ax, Get16(m->system->real + 0x413));
 }
 
+static void OnPrinterService(void) {
+  m->ah = 0xb0;  // "no printer": not busy, out of paper, selected
+}
+
+static void OnTimeServiceGetSystemTime(void) {
+  u64 DAY_SECS = 24UL * 60 * 60;
+  struct timespec now = GetTime();
+  u64 currday, daytime;
+  u8 midnights = 0;
+  unassert(now.tv_sec >= DAY_SECS);
+  currday = (u64)now.tv_sec / DAY_SECS;
+  // calculate the number of midnights elapsed since the last call here
+  // this will also reset the midnight count
+  if (prevday != 0) {
+    if (currday > prevday) midnights = currday - prevday;
+  }
+  prevday = currday;
+  // calculate nanoseconds from day start
+  daytime = (u64)now.tv_sec - currday * DAY_SECS;
+  daytime = daytime * 1000000000L + now.tv_nsec;
+  // calculate BIOS system timer ticks from day start
+  daytime = daytime * (0x1800B0L / 80) / (DAY_SECS * 1000000000L / 80);
```

Comment:
Hello @tkchia,

I am reviewing the INT 1A time services code for possible inclusion into `blink16`, but I don't quite understand the divide by 80 twice in the above conversion. I assume its to convert back from the multiplication by `DAY_SECS` earlier. Shouldn't the two `80`s be `60`s?

Thank you!

Reply:
@ghaerr : 80 is the greatest common divisor of `0x1800B0` and 24 &times; 60 &times; 60 &times; 10⁹.  Basically I want to convert a nanosecond count into a BIOS timer tick count, and there are `0x1800B0` timer ticks in a day.  Thank you!

Reply:
@ghaerr : a na&iuml;ve way to do the conversion would be to say
```
  daytime = daytime * 0x1800B0 / (DAY_SECS * 1000000000L);
```
but this is susceptible to 64-bit arithmetic overflow.  Removing the common factor of 80 from the multiplier and the divisor also happens to avoid this overflow.  Thank you!